src/vectorpostprocessors/Fragility.C

```diff
@@ -0,0 +1,208 @@
+               "for median and lognormal standard deviation of the "
+               "component.");
+  if (_median_fragility_limits[0] >= _median_fragility_limits[1])
+    mooseError("Error in block '" + name() +
+               "'. Minimum median fragility should be less than the maximum median fragility.");
+  if (_beta_fragility_limits[0] >= _beta_fragility_limits[1])
+    mooseError(
+        "Error in block '" + name() +
+        "'. Minimum should be less than the maximum for fragility lognormal standard deviation.");
+  if (_median_fragility_limits[0] <= 0 || _beta_fragility_limits[0] <= 0)
+    mooseError("Error in block '" + name() +
+               "'. Limits of median and beta of the component fragility must be positive.");
+  // Check if _im_values is of the same size as the number of bins
+  if (_im_values.size() != _num_bins)
+    mooseError("Error in block '" + name() +
+               "'. Number of IM values should be the same as the number of bins.");
+}
+
+void
+Fragility::initialize()
+{
+  _median_demand.clear();
+  _beta_demand.clear();
+  _im.clear();
+  _conditional_pf.clear();
+  _median_fragility.clear();
+  _beta_fragility.clear();
+}
+
+void
+Fragility::execute()
+{
+  _im.resize(_num_bins);
+  _median_demand.resize(_num_bins);
+  _beta_demand.resize(_num_bins);
+  _conditional_pf.resize(_num_bins);
+  _median_fragility.resize(1);
+  _beta_fragility.resize(1);
+  for (std::size_t bin = 0; bin < _num_bins; bin++)
+  {
+    _im[bin] = _im_values[bin];
+    std::vector<Real> stoc_demands = Fragility::calcDemandsFromFile(
+        bin); // calculating demands from csv files if it is a restart PRA
+    _median_demand[bin] = MastodonUtils::median(stoc_demands);
+    _beta_demand[bin] = MastodonUtils::lognormalStandardDeviation(stoc_demands);
+    boost::math::lognormal_distribution<Real> demand_distribution(log(_median_demand[bin]),
+                                                                  _beta_demand[bin]);
+    boost::math::lognormal_distribution<Real> capacity_distribution(log(_median_cap),
+                                                                    _beta_ssc_cap);
+    _conditional_pf[bin] = MastodonUtils::greaterProbability<boost::math::lognormal>(
+        demand_distribution, capacity_distribution);
+    _console << "**********\nFinished calculations for bin: " << bin
+             << " \n Median demand: " << _median_demand[bin]
+             << " \n Lognormal standard deviation: " << _beta_demand[bin]
+             << " \n Conditional probability of failure: " << _conditional_pf[bin]
+             << "\n**********\n";
+  }
+  std::vector<Real> fitted_vals = MastodonUtils::maximizeLogLikelihood(
+      _im, _conditional_pf, _median_fragility_limits, _beta_fragility_limits, 500);
+  _median_fragility[0] = fitted_vals[0];
+  _beta_fragility[0] = fitted_vals[1];
+}
+
+std::vector<Real>
+Fragility::calcDemandsFromFile(unsigned int bin)
+{
+  std::vector<Real> demand_sample;
```

Comment:
Probably should just wrap this in BOOST stuff as well, make it return an empty vector if BOOST is not setup.